scripts/eqtl_hail_batch/conditional_analysis.py

```diff
@@ -0,0 +1,459 @@
+#!/usr/bin/env python3
+
+"""Perform conditional analysis on SNPs and expression residuals"""
+
+import os
+import hail as hl
+import hailtop.batch as hb
+import pandas as pd
+import statsmodels.api as sm
+from patsy import dmatrices  # pylint: disable=no-name-in-module
+from scipy.stats import spearmanr
+from cpg_utils.hail_batch import (
+    dataset_path,
+    output_path,
+    copy_common_env,
+    init_batch,
+    remote_tmpdir,
+)
+from cloudpathlib import AnyPath
+import click
+from multipy.fdr import qvalue
+
+
+DEFAULT_DRIVER_MEMORY = '4G'
+DRIVER_IMAGE = 'australia-southeast1-docker.pkg.dev/cpg-common/images/multipy:0.16'
+assert DRIVER_IMAGE
+
+TOB_WGS = dataset_path('mt/v7.mt/')
+FREQ_TABLE = dataset_path('joint-calling/v7/variant_qc/frequencies.ht/', 'analysis')
+
+
+def get_number_of_scatters(residual_df: pd.DataFrame, significant_snps_df: pd.DataFrame) -> int:
+    """get index of total number of genes"""
+
+    # Identify the top eSNP for each eGene and assign remaining to df
+    esnp1 = (
+        significant_snps_df.sort_values(['gene_symbol', 'fdr'], ascending=True)
+        .groupby('gene_symbol')
+        .first()
+        .reset_index()
+    )
+    gene_ids = esnp1['gene_symbol'][esnp1['gene_symbol'].isin(residual_df.columns)]
+
+    return len(gene_ids)
+
+
+def prepare_genotype_info(keys_path):
+
+    init_batch()
+    filtered_mt_path = output_path('genotype_table.ht', 'tmp')
+    if not hl.hadoop_exists(filtered_mt_path):
+        mt = hl.read_matrix_table(TOB_WGS)
+        mt = hl.experimental.densify(mt)
+        # filter to biallelic loci only
+        mt = mt.filter_rows(hl.len(mt.alleles) == 2)
+        # filter out variants that didn't pass the VQSR filter
+        mt = mt.filter_rows(hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)
+        # VQSR does not filter out low quality genotypes. Filter these out
+        mt = mt.filter_entries(mt.GQ <= 20, keep=False)
+        # filter out samples with a genotype call rate > 0.8 (as in the gnomAD supplementary paper)
+        n_samples = mt.count_cols()
+        call_rate = 0.8
+        mt = mt.filter_rows(
+            hl.agg.sum(hl.is_missing(mt.GT)) > (n_samples * call_rate), keep=False
+        )
+        # filter out variants with MAF < 0.01
+        ht = hl.read_table(FREQ_TABLE)
+        mt = mt.annotate_rows(freq=ht[mt.row_key].freq)
+        mt = mt.filter_rows(mt.freq.AF[1] > 0.01)
```

Comment:
Should this perhaps be >= 0.01 - here we keep anything with a MAF > 0.01, and dropping everything MAF <= 0.01?

Reply:
Sorry, I'm a little confused- when you say "dropping everything MAF <= 0.01", isn't that what it's currently doing?

Reply:
Sorry, that was a _terribly_ written comment from me! Yes, filtering out MAF <= 0.01 is what the code is currently doing - but your comment says we're filtering out MAF < 0.01.

So this is being super nit picky, but just to make sure the description of what we do matches up with what we actually do (and what Seyhan did) so that it's easy to write up an accurate methods description later.

Reply:
makes sense. Comments in both scripts should now reflect that I'm filtering in this way (so there's no confusion later)